aptos-move/framework/aptos-framework/sources/account/common_account_abstractions/ethereum_derivable_account.move

```diff
@@ -0,0 +1,335 @@
+/// Derivable account abstraction that verifies a message signed by
+/// SIWE.
+/// 1. The message format is as follows:
+///
+/// <domain> wants you to sign in with your Ethereum account:
+/// <ethereum_address>
+///
+/// Please confirm you explicitly initiated this request from <domain>. You are approving to execute transaction <entry_function_name> on Aptos blockchain (<network_name>).
+///
+/// URI: <domain>
+/// Version: 1
+/// Chain ID: <chain_id>
+/// Nonce: <digest>
+/// Issued At: <issued_at>
+///
+/// 2. The abstract public key is a BCS serialized `SIWEAbstractPublicKey`.
+/// 3. The abstract signature is a BCS serialized `SIWEAbstractSignature`.
+/// 4. This module has been tested for the following wallets:
+/// - Metamask
+/// - Phantom
+/// - Coinbase
+/// - OKX
+/// - Exodus
+/// - Backpack
+
+module aptos_framework::ethereum_derivable_account {
+    use aptos_framework::auth_data::AbstractionAuthData;
+    use aptos_framework::common_account_abstractions_utils::{network_name, entry_function_name};
+    use aptos_std::secp256k1;
+    use aptos_std::option;
+    use aptos_std::aptos_hash;
+    use aptos_std::base16::base16_utf8_to_vec_u8;
+    use std::bcs_stream::{Self, deserialize_u8};
+    use std::chain_id;
+    use std::string_utils;
+    use std::transaction_context;
+    use std::vector;
+    use std::string::{Self, String};
+
+    /// Signature failed to verify.
+    const EINVALID_SIGNATURE: u64 = 1;
+    /// Entry function payload is missing.
+    const EMISSING_ENTRY_FUNCTION_PAYLOAD: u64 = 2;
+    /// Invalid signature type.
+    const EINVALID_SIGNATURE_TYPE: u64 = 3;
+    /// Address mismatch.
+    const EADDR_MISMATCH: u64 = 4;
+    /// Unexpected v value.
+    const EUNEXPECTED_V: u64 = 5;
+
+    enum SIWEAbstractSignature has drop {
+        EIP1193DerivedSignature {
+            issued_at: String,
+            signature: vector<u8>,
+        },
+    }
+
+    struct SIWEAbstractPublicKey has drop {
+        // The Ethereum address, with 0x prefix, in utf8 bytes
+        ethereum_address: vector<u8>,
```

Comment:
I am wondering if it is better to use here the full public key (i.e. not just last 20 bytes), and then derive ethereum address from it to put in the message.

@alinush for opinions , and if secp256k1 validation via secp256k1::ecdsa_recover to get pub_key, cut it to address and check address is the right way to verify ethereum signatures

Reply:
I don't know if you'll ever be able to "get your hands on" the Ethereum public key though, cause they (I think) always verify signatures via [pubkey recovery](https://alinush.github.io/ecdsa#pubkey-recovery) so the pubkeys never end up on-chain.

So not sure how exposed it is. e.g., does the wallet even return it with the ECDSA signature when it signs a `signMessage`/SIWE prompt?

In any case, it would be better to have the pubkey here though, is my hunch.

It would also be better to do a normal ECDSA signature verification (even if you don't end up using the pubkey here).

Reply:
> Could you clarify? are you suggesting to do something else than what we are doing now?

Ah, nevermind: the only thing you can do is a pubkey-recovery-based signature verification because Ethereum wallets will not give you a pubkey...

So you are 👌

Reply:
It's very peculiar that you are not storing the address bytes here, but some base16 (or whatever) UTF-8 encoding of them...

Why?

Reply:
same as we do in Solana https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/account/common_account_abstractions/solana_derivable_account.move#L59